### PR TITLE
Pruning with solver queries

### PR DESCRIPTION
**Pull request by manasij7479**
Has working pruning implementation with the KLEE backend and a WIP version with Alive2.
Also renames DataflowPruning to Pruning

### CODE REVIEW

File: include/souper/Infer/Pruning.h [hunk 1]
```diff
@@ -2,51 +2,42 @@
 
 #include "souper/Infer/Interpreter.h"
 #include "souper/Inst/Inst.h"
+#include "souper/Extractor/Solver.h"
```
Comment (regehr) [inline include/souper/Infer/Pruning.h]: alphabetical please

File: include/souper/Infer/AliveDriver.h [hunk 2]
```diff
@@ -63,6 +64,9 @@ class AliveDriver {
 bool isTransformationValid(Inst* LHS, Inst* RHS, const std::vector<InstMapping> &PCs,
                            InstContext &IC);
 
+bool isCandidateInfeasible(Inst *RHS, ValueCache &C, llvm::APInt LHSValue,
+                          InstContext &IC);
```
Comment (regehr) [inline include/souper/Infer/AliveDriver.h]: space/tab problem? the two capital "I" characters should line up

File: include/souper/Infer/Pruning.h [hunk 3]
```diff
@@ -2,51 +2,42 @@
...
-
 typedef std::function<bool(Inst *, std::vector<Inst *> &)> PruneFunc;
 
-class DataflowPruningManager {
+class PruningManager {
 public:
-  DataflowPruningManager(Inst *LHS_, std::vector<Inst *> &Inputs_,
-                         unsigned StatsLevel);
+  PruningManager(Inst* LHS_, std::vector< souper::Inst* >& Inputs_,
+                         unsigned int StatsLevel_, InstContext& IC_,
```
Comment (regehr) [inline include/souper/Infer/Pruning.h]: line these up please

File: include/souper/Infer/Pruning.h [hunk 4]
```diff
@@ -2,51 +2,42 @@
...
-
-
 typedef std::function<bool(Inst *, std::vector<Inst *> &)> PruneFunc;
 
-class DataflowPruningManager {
+class PruningManager {
 public:
-  DataflowPruningManager(Inst *LHS_, std::vector<Inst *> &Inputs_,
-                         unsigned StatsLevel);
+  PruningManager(Inst* LHS_, std::vector< souper::Inst* >& Inputs_,
```
Comment (regehr) [inline include/souper/Infer/Pruning.h]: always put the "*" or "&" next to the name, not next to the type

**Comment by regehr** [comment]
is this ready to go?

**Comment by regehr** [comment]
cool.
can this be tested?

**Comment by manasij7479** [comment]
Yes
Need to check for the string "pruned using Solver!"
The policy for when to enable solver pruning is unclear at this moment, so adding tests immediately might not be the best idea.

**Comment by regehr** [comment]
ok I'll merge (after reading again) with the understanding that you WILL do the test later, ok?

**Comment by manasij7479** [comment]
Sounds good

**Comment by regehr** [comment]
LGTM , thanks!

**Pull request merged**